pkg/k8s/service_cache.go

```diff
@@ -369,3 +377,118 @@ func (s *ServiceCache) UniqueServiceFrontends() map[string]struct{} {
+
+	localEndpoints, hasLocalEndpoints := s.endpoints[id]
+	if hasLocalEndpoints {
+		for ip, e := range localEndpoints.Backends {
+			endpoints.Backends[ip] = e
+		}
+	}
+
+	svc, hasExternalService := s.services[id]
+	if hasExternalService && svc.IncludeExternal {
+		externalEndpoints, hasExternalEndpoints := s.externalEndpoints[id]
+		if hasExternalEndpoints {
+			for clusterName, remoteClusterEndpoints := range externalEndpoints.endpoints {
+				if clusterName == option.Config.ClusterName {
+					continue
+				}
+
+				for ip, e := range remoteClusterEndpoints.Backends {
+					if _, ok := endpoints.Backends[ip]; ok {
+						log.WithFields(logrus.Fields{
+							logfields.K8sSvcName:   id.Name,
+							logfields.K8sNamespace: id.Namespace,
+							logfields.IPAddr:       ip,
+							"cluster":              clusterName,
+						}).Warning("Conflicting service backend IP")
+					} else {
+						endpoints.Backends[ip] = e
+					}
+				}
+			}
+		}
+	}
+
+	// If the no local endpoints is not present yet and not remote
+	// endpoints are defined, return nil to indicate the service is not
+	// ready yet.
+	if !hasLocalEndpoints && len(endpoints.Backends) == 0 {
+		return nil
+	}
+
+	return endpoints
+}
+
+// MergeExternalServiceUpdate merges a cluster service of a remote cluster into
+// the local service cache. The service endpoints are stored as external endpoints
+// and are correlated on demand with local services via correlateEndpoints().
+func (s *ServiceCache) MergeExternalServiceUpdate(service *service.ClusterService) {
+	id := ServiceID{Name: service.Name, Namespace: service.Namespace}
+
+	// Ignore updates of own cluster
+	if service.Cluster == option.Config.ClusterName {
+		return
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	externalEndpoints, ok := s.externalEndpoints[id]
+	if !ok {
+		externalEndpoints = newExternalEndpoints()
+		s.externalEndpoints[id] = externalEndpoints
+	}
+
+	externalEndpoints.endpoints[service.Cluster] = &Endpoints{
+		Backends: service.Backends,
+	}
+
+	svc, ok := s.services[id]
+	if ok {
+		s.Events <- ServiceEvent{
+			Action:    UpdateService,
+			ID:        id,
+			Service:   svc,
+			Endpoints: s.correlateEndpoints(id),
+		}
+	}
+}
+
+// MergeExternalServiceDelete merges the deletion a cluster service in a remote
+// cluster into the local service cache. The service endpoints are stored as
+// external endpoints and are correlated on demand with local services via
+// correlateEndpoints().
+func (s *ServiceCache) MergeExternalServiceDelete(service *service.ClusterService) {
+	id := ServiceID{Name: service.Name, Namespace: service.Namespace}
+
+	// Ignore updates of own cluster
+	if service.Cluster == option.Config.ClusterName {
+		return
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	externalEndpoints, ok := s.externalEndpoints[id]
+	if ok {
+
+		delete(externalEndpoints.endpoints, service.Cluster)
+
+		svc, ok := s.services[id]
+		if ok {
+			s.Events <- ServiceEvent{
+				Action:    UpdateService,
```

Comment:
Shouldn't this be a `DeleteService`?

Reply:
The correct logic is this:
```
                          event := ServiceEvent{
                                  Action:    UpdateService,
                                  ID:        id,
                                  Service:   svc,
                                  Endpoints: endpoints,
                          }

                          if !serviceReady {
                                  event.Action = DeleteService
                          }

                          s.Events <- event
```

It's in the latest pushed code